### DISCUSSION TITLE
Make the value of one field depend on another

### DISCUSSION THREAD

**Discussion by Nick-Mazuk** [colinhacks/zod#3172] [Q&A] [answered]
Is it possible to ensure the potential values of one field depend on the value of another field? If so, is it possible to add the issue to the field instead of the overall object?

Here's a more concrete use case. I want to validate a form using [svelte-superforms](https://superforms.rocks/) where two different input values depend on each other. And this has two subproblems:

- Validating the overall form inputs.
- Showing the error on the fields, not for the entire form.

## Validating the form

This form is for a product listing. The minimum price for the product depends on it's category. So you could imagine having a schema for the form that looks something like below:

```ts
z.object({
    category: z.string(),
    priceCents: z.number(),
})
```

And the price minimums would look something like the following:

| Category | Price minimum (cents) |
| -------- | --------------------- |
| A        | 4999                  |
| B        | 499                   |
| C        | 999                   |

I want to make sure that any combination of category and price is correct. For instance, a product with a price of 1500 is valid in categories B and C, but not category A. A product with a price of 10000 is valid in all categories. A product with price 1 is not valid in any category.

Guessing `.superRefine` would solve this?

## Showing the error message inline

The optimal UX for my use case is to show any errors in line with the price field—not on the overall form during submission. That way sellers can know to change the price for their category.

The way that svelte-superforms works is that if an issue is raised on a field in the object, it will automatically show it on the corresponding input.

For instance, if I the price did not depend on the category, I could have the following schema. Inputting a price of 1 will show the error on the price input.

```ts
z.object({
    category: z.string(),
    priceCents: z.number().gte(4999),
})
```

I'd like to have that same UX but for when the minimum price depends on the category.

So if `.superRefine` is the correct way to detect an invalid schema, how can I raise an issue on the `priceCents` field rather than the object as a whole?

## Related discussions

Fields depending on other fields: https://github.com/colinhacks/zod/discussions/954 – The solution described here doesn't best suit my use case. It seems to work well for just a few predefined categories, but I have hundreds of categories and nearing the thousands of categories. I'm concerned with performance as things grow since it's O(n). But I can calculate the price minimum in O(1).

**Comment by JacobWeisenburger** [comment]
Perhaps this would help:
https://github.com/colinhacks/zod#discriminated-unions

```ts
const schema = z.discriminatedUnion( 'category', [
    z.object( {
        category: z.literal( 'A' ),
        priceCents: z.number().min( 4999 ),
    } ),
    z.object( {
        category: z.literal( 'B' ),
        priceCents: z.number().min( 499 ),
    } ),
    z.object( {
        category: z.literal( 'C' ),
        priceCents: z.number().min( 999 ),
    } ),
] )
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by Nick-Mazuk** [reply]
Hey Jacob, as I mentioned I'm concerned with the performance of that approach. In reality, I have hundreds of categories nearing into the thousands.

For any given category literal, I can determine the minimum price in O(1). But my understanding is that using a Zod `discriminatedUnion` results in O(n) performance.

**Comment by JacobWeisenburger** [comment, marked as answer]
Would something like this work better for you?

```ts
const minLookup = {
    A: 4999,
    B: 499,
    C: 999,
}

type Keys = keyof typeof minLookup

const schema = z.object( {
    category: z.enum( Object.keys( minLookup ) as [ Keys, ...Keys[] ] ),
    priceCents: z.number(),
} ).superRefine( ( data, ctx ) => {
    const min = minLookup[ data.category ]
    if ( data.priceCents < min ) {
        ctx.addIssue( {
            code: z.ZodIssueCode.custom,
            message: `priceCents must be at least ${ min }`,
            path: [ 'priceCents' ],
        } )
    }
} )
type Data = z.infer<typeof schema>
// type Data = {
//     category: "A" | "B" | "C";
//     priceCents: number;
// }

console.log( schema.safeParse( {
    category: 'A',
    priceCents: 5000,
} ).success ) // true

const result = schema.safeParse( {
    category: 'A',
    priceCents: 4242,
} )
!result.success && console.log( result.error.issues )
// [
//     {
//         code: "custom",
//         message: "priceCents must be at least 4999",
//         path: [ "priceCents" ],
//     }
// ]
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
